include/zephyr/sys/barrier_builtin.h

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright (c) 2023 Carlo Caione <ccaione@baylibre.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_SYS_BARRIER_BUILTIN_H_
+#define ZEPHYR_INCLUDE_SYS_BARRIER_BUILTIN_H_
+
+#ifndef ZEPHYR_INCLUDE_SYS_BARRIER_H_
+#error Please include <zephyr/sys/barrier.h>
+#endif
+
+#include <zephyr/toolchain.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+static ALWAYS_INLINE void z_barrier_dmem_fence_full(void)
+{
+#if defined(__STDC_VERSION__) && __STDC_VERSION__ < 201112L
+	/* C99*/
+	__atomic_thread_fence(__ATOMIC_SEQ_CST);
```

Comment:
This really has nothing to do with C99. Since `__atomic_thread_fence` is a GCC built-in/GNU extension, it should check for `__GNUC__` here, and otherwise assume the C11 stdatomic.

```suggestion
#if defined(__GNUC__)
	/* C99*/
	__atomic_thread_fence(__ATOMIC_SEQ_CST);
```